File: multiple-3-or-5/multiple-3-or-5.pl
prog_sum(N, Sum) :- 
    Sum is (N * (N+1)) div 2.

amount_of_divisions_in_range(Number, N, Am) :-
    (Number mod N =:= 0 -> T is 1; T is 0),
    Am is (Number div N) - T.

solution(Number, Result) :- 
    amount_of_divisions_in_range(Number, 3, N3),
    amount_of_divisions_in_range(Number, 5, N5),
    amount_of_divisions_in_range(Number, 15, N15),
    prog_sum(N3, Sum3),
    prog_sum(N5, Sum5),
    prog_sum(N15, Sum15),
    Result is 3 * Sum3 + 5 * Sum5 - 15 * Sum15.

File: multiple-3-or-5/multiple-3-or-5.rs
fn prog_sum(n: i32) -> i32 {
    n * (n + 1) / 2
}

fn amount_of_divisions_in_range(number: i32, n: i32) -> i32 {
    number / n - if number % n == 0 { 1 } else { 0 }
}

fn solution(num: i32) -> i32 {
    if num > 0 {
        3 * prog_sum(amount_of_divisions_in_range(num, 3))
            + 5 * prog_sum(amount_of_divisions_in_range(num, 5))
            - 15 * prog_sum(amount_of_divisions_in_range(num, 15))
    } else {
        0
    }
}
